File: 221.py
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        m = len(matrix)
        n = len(matrix[0])
        memo = [1 if matrix[0][j] == '1' else 0 for j in range(n)]
        maxside = max(memo)
        for i in range(1, m):
            leftup = memo[0]
            memo[0] = 1 if matrix[i][0] == '1' else 0
            maxside = max(maxside, memo[0])
            for j in range(1, n):
                leftup, memo[j] = memo[j], (1 + min(leftup, memo[j - 1], memo[j]) if matrix[i][j] == '1' else 0)
                maxside = max(maxside, memo[j])
        return maxside * maxside

File: sol/221/221.cpp
struct Solution {
    int maximalSquare(const vector<vector<char>>& matrix) {
        int m = matrix.size(), n = matrix.front().size(), maxside = 0, leftup, temp;
        vector<int> memo(n, 0);
        for (int j = 0; j < n; ++j){
            memo[j] = (matrix[0][j] == '1' ? 1 : 0);
            maxside = max(maxside, memo[j]);
        }
        for (int i = 1; i < m; ++i){
            leftup = memo[0];
            memo[0] = (matrix[i][0] == '1' ? 1 : 0);
            maxside = max(maxside, memo[0]);
            for (int j = 1; j < n; ++j){
                temp = leftup;
                leftup = memo[j];
                memo[j] = (matrix[i][j] == '1' ? min(min(memo[j - 1], memo[j]), temp) + 1 : 0);
                maxside = max(maxside, memo[j]);
            }
        }
        return maxside * maxside;
    }
};
